File: apps/windows/TeslaSync.App/widget-primitives/WidgetDetailCard.Model.cs
using System.Globalization;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.WidgetPrimitives.WidgetDetailCardSurface;

/// <summary>
/// Canonical metadata for the <c>WidgetDetailCard</c> widget primitive — the native mirror of the web
/// component at <c>web/src/features/dashboard/widgets/shared/WidgetDetailCard.tsx</c>: the stable
/// diagnostics slug. UI-free so the metadata is asserted in tests.
/// </summary>
public static class WidgetDetailCardRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11).</summary>
    public const string Slug = "WidgetDetailCard";
}

/// <summary>
/// The semantic badge tone — the native mirror of the web <c>DetailEntry.badge.variant</c> union
/// (<c>'success' | 'warning' | 'error' | 'neutral'</c> in the web source). The web maps these to the shared
/// <c>Badge</c> variants via its <c>badgeVariantMap</c> (success → success, warning → warning, error →
/// danger, neutral → neutral); <see cref="WidgetDetailCardProjection.StatusFor"/> reproduces that mapping onto
/// the native <see cref="StatusKind"/>.
/// </summary>
public enum WidgetDetailBadgeVariant
{
    /// <summary>web <c>'success'</c> — maps to <see cref="StatusKind.Success"/>.</summary>
    Success,

    /// <summary>web <c>'warning'</c> — maps to <see cref="StatusKind.Warning"/>.</summary>
    Warning,

    /// <summary>web <c>'error'</c> — maps to the danger tone (<see cref="StatusKind.Danger"/>).</summary>
    Error,

    /// <summary>web <c>'neutral'</c> — maps to <see cref="StatusKind.Neutral"/>.</summary>
    Neutral,
}

/// <summary>
/// A trailing chip on a detail row — the native analogue of the web <c>DetailEntry.badge</c>
/// (<c>{ text, variant }</c>): the short <see cref="Text"/> and its semantic <see cref="Variant"/>.
/// </summary>
/// <param name="Text">The chip label (web <c>badge.text</c>).</param>
/// <param name="Variant">The semantic tone (web <c>badge.variant</c>).</param>
public sealed record WidgetDetailBadge(string Text, WidgetDetailBadgeVariant Variant);

/// <summary>
/// One label / value row — the native analogue of the web <c>DetailEntry</c> interface
/// (<c>label</c>, <c>value: string | number | null</c>, optional <c>badge</c>, optional <c>mono</c>). The web
/// renders <c>value ?? '—'</c>, so a null value (or a null number) becomes the muted em dash at projection
/// time. The two web value styles are reproduced by the <see cref="Text"/> and <see cref="Number"/> factories;
/// pure data (no WinUI types) so it is exercised headlessly.
/// </summary>
public sealed record WidgetDetailEntry
{
    private WidgetDetailEntry(string label, string? value, WidgetDetailBadge? badge, bool mono)
    {
        Label = label;
        Value = value;
        Badge = badge;
        Mono = mono;
    }

    /// <summary>The row label (web <c>label</c>) — shown left, uppercased, muted.</summary>
    public string Label { get; }

    /// <summary>The already-resolved display value (web <c>value</c>); null renders the muted em dash.</summary>
    public string? Value { get; }

    /// <summary>Optional trailing chip (web <c>badge</c>); null renders no chip.</summary>
    public WidgetDetailBadge? Badge { get; }

    /// <summary>When true the value uses the monospace family (web <c>mono</c> → <c>font-mono</c>).</summary>
    public bool Mono { get; }

    /// <summary>
    /// A string-valued row (web <c>value: string | null</c>): the value is shown verbatim, or the muted em
    /// dash when <paramref name="value"/> is null.
    /// </summary>
    /// <param name="label">The row label (web <c>label</c>).</param>
    /// <param name="value">The display value (web <c>value</c>); null renders the muted em dash.</param>
    /// <param name="badge">Optional trailing chip (web <c>badge</c>).</param>
    /// <param name="mono">When true the value uses the monospace family (web <c>mono</c>).</param>
    public static WidgetDetailEntry Text(
        string label,
        string? value,
        WidgetDetailBadge? badge = null,
        bool mono = false) =>
        new(label, value, badge, mono);

    /// <summary>
    /// A number-valued row (web <c>value: number | null</c>): a non-null number is rendered with the invariant
    /// culture (matching the web's <c>String(number)</c>); a null renders the muted em dash.
    /// </summary>
    /// <param name="label">The row label (web <c>label</c>).</param>
    /// <param name="value">The numeric value (web <c>value</c>); null renders the muted em dash.</param>
    /// <param name="badge">Optional trailing chip (web <c>badge</c>).</param>
    /// <param name="mono">When true the value uses the monospace family (web <c>mono</c>).</param>
    public static WidgetDetailEntry Number(
        string label,
        double? value,
        WidgetDetailBadge? badge = null,
        bool mono = false) =>
        new(label, value?.ToString(CultureInfo.InvariantCulture), badge, mono);
}

/// <summary>
/// The render-time data model the <c>WidgetDetailCard</c> view binds to — the native analogue of the web
/// <c>WidgetDetailCardProps</c> (<c>entries</c>, <c>compact</c>, <c>emptyMessage</c>, <c>emptyIcon</c>). The web
/// component is purely presentational: its parent widget owns any data fetching and feeds an already-resolved
/// list of rows, so — exactly like React re-rendering with resolved props — there is no fetch-driven loading /
/// error / stale / offline branch to reproduce here; the only branches are "empty" (no rows → the friendly
/// empty surface) and "populated" (the rows, sliced to the first four in <see cref="Compact"/> mode). Pure data
/// (no WinUI types) so the projection is unit-tested without a UI host.
/// </summary>
public sealed record WidgetDetailCardModel
{
    private WidgetDetailCardModel(
        IReadOnlyList<WidgetDetailEntry> entries,
        bool compact,
        string? emptyMessage,
        string? emptyIconGlyph)
    {
        Entries = entries;
        Compact = compact;
        EmptyMessage = emptyMessage;
        EmptyIconGlyph = emptyIconGlyph;
    }

    /// <summary>The rows to render (web <c>entries</c>); an empty list renders the empty surface.</summary>
    public IReadOnlyList<WidgetDetailEntry> Entries { get; }

    /// <summary>When true only the first four rows are shown (web <c>compact</c> → <c>entries.slice(0, 4)</c>).</summary>
    public bool Compact { get; }

    /// <summary>Optional caller override for the empty message (web <c>emptyMessage</c>); null uses the localized default.</summary>
    public string? EmptyMessage { get; }

    /// <summary>Optional Segoe Fluent glyph for the empty surface (web <c>emptyIcon</c>); null shows no glyph.</summary>
    public string? EmptyIconGlyph { get; }

    /// <summary>The initial / no-data model — an empty row list, rendering the empty surface.</summary>
    public static WidgetDetailCardModel Empty { get; } = Create(Array.Empty<WidgetDetailEntry>());

    /// <summary>Build a model over a row list and the optional web props.</summary>
    /// <param name="entries">The rows to render (web <c>entries</c>).</param>
    /// <param name="compact">When true only the first four rows are shown (web <c>compact</c>).</param>
    /// <param name="emptyMessage">Optional empty-message override (web <c>emptyMessage</c>).</param>
    /// <param name="emptyIconGlyph">Optional empty-surface glyph (web <c>emptyIcon</c>).</param>
    public static WidgetDetailCardModel Create(
        IReadOnlyList<WidgetDetailEntry> entries,
        bool compact = false,
        string? emptyMessage = null,
        string? emptyIconGlyph = null)
    {
        ArgumentNullException.ThrowIfNull(entries);
        return new WidgetDetailCardModel(entries, compact, emptyMessage, emptyIconGlyph);
    }
}

/// <summary>
/// A fully projected, render-ready row — the native analogue of one mapped <c>&lt;div&gt;</c> in the web
/// source: the original <see cref="Label"/> (kept for the accessible name; the view uppercases the glyph), the
/// resolved <see cref="DisplayValue"/> (web <c>value ?? '—'</c>), the <see cref="Mono"/> flag, the resolved
/// badge (<see cref="HasBadge"/> / <see cref="BadgeText"/> / <see cref="BadgeStatus"/>), the
/// <see cref="ShowDivider"/> hairline guard (web <c>i &lt; visible.length - 1</c>) and the composed
/// <see cref="AutomationName"/> Narrator reads.
/// </summary>
/// <param name="Label">The original row label (web <c>label</c>).</param>
/// <param name="DisplayValue">The resolved value, or the muted em dash (web <c>value ?? '—'</c>).</param>
/// <param name="Mono">True when the value uses the monospace family (web <c>mono</c>).</param>
/// <param name="HasBadge">True when a trailing chip is shown (web <c>entry.badge</c> present).</param>
/// <param name="BadgeText">The chip label (web <c>badge.text</c>); empty when <see cref="HasBadge"/> is false.</param>
/// <param name="BadgeStatus">The chip tone mapped onto <see cref="StatusKind"/> (web <c>badgeVariantMap</c>).</param>
/// <param name="ShowDivider">True for every row but the last (web bottom-border guard).</param>
/// <param name="AutomationName">The composed accessible name Narrator reads for the row.</param>
public sealed record WidgetDetailRowDisplay(
    string Label,
    string DisplayValue,
    bool Mono,
    bool HasBadge,
    string BadgeText,
    StatusKind BadgeStatus,
    bool ShowDivider,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready view of a <see cref="WidgetDetailCardModel"/> — everything the web source
/// derives before returning JSX: the <see cref="IsEmpty"/> branch guard (web <c>entries.length === 0</c>), the
/// resolved <see cref="EmptyMessage"/> and optional <see cref="EmptyIconGlyph"/>, and the visible
/// <see cref="Rows"/> (sliced to four in compact mode). Pure data so every value is asserted headlessly.
/// </summary>
/// <param name="IsEmpty">True when there are no rows — the friendly empty surface renders (web empty branch).</param>
/// <param name="EmptyMessage">The resolved empty message (web <c>emptyMessage ?? 'No details available'</c>).</param>
/// <param name="EmptyIconGlyph">Optional empty-surface glyph (web <c>emptyIcon</c>); null shows no glyph.</param>
/// <param name="Rows">The visible rows; empty when <see cref="IsEmpty"/> is true.</param>
public sealed record WidgetDetailCardDisplay(
    bool IsEmpty,
    string EmptyMessage,
    string? EmptyIconGlyph,
    IReadOnlyList<WidgetDetailRowDisplay> Rows);

/// <summary>
/// Pure projection from a <see cref="WidgetDetailCardModel"/> to its <see cref="WidgetDetailCardDisplay"/> — the
/// native port of <c>web/src/features/dashboard/widgets/shared/WidgetDetailCard.tsx</c>. Reproduces the web
/// derivations exactly:
/// <list type="bullet">
///   <item><description>an empty row list yields the empty surface (web <c>entries.length === 0</c>), whose
///   message is the caller override or the localized <c>widget.detail.empty</c> default (web
///   <c>emptyMessage ?? 'No details available'</c>) — the surface always renders rather than collapsing.</description></item>
///   <item><description>compact mode shows only the first <see cref="CompactLimit"/> rows (web
///   <c>entries.slice(0, 4)</c>).</description></item>
///   <item><description>each row resolves <c>value ?? '—'</c>, maps the badge variant through
///   <see cref="StatusFor"/> (web <c>badgeVariantMap</c>), and carries a bottom-hairline guard for every row but
///   the last (web <c>i &lt; visible.length - 1</c>).</description></item>
/// </list>
/// No WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
public static class WidgetDetailCardProjection
{
    /// <summary>The compact-mode row cap (web <c>entries.slice(0, 4)</c>).</summary>
    public const int CompactLimit = 4;

    /// <summary>The muted no-data glyph (web <c>'—'</c>).</summary>
    public const string EmDash = "\u2014";

    /// <summary>i18n key for the default empty message (the web literal is keyed for translation here).</summary>
    public const string EmptyMessageKey = "widget.detail.empty";

    /// <summary>English fallback for <see cref="EmptyMessageKey"/> (the web default literal).</summary>
    public const string EmptyMessageFallback = "No details available";

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web props).</param>
    /// <param name="localizer">The i18n facade the empty message resolves through (P1/S10).</param>
    /// <returns>The render-ready display model.</returns>
    public static WidgetDetailCardDisplay Project(WidgetDetailCardModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        // web: if (entries.length === 0) return <EmptyState ... />
        if (model.Entries.Count == 0)
        {
            string message = model.EmptyMessage ?? localizer.GetString(EmptyMessageKey, EmptyMessageFallback);
            return new WidgetDetailCardDisplay(
                IsEmpty: true,
                EmptyMessage: message,
                EmptyIconGlyph: model.EmptyIconGlyph,
                Rows: Array.Empty<WidgetDetailRowDisplay>());
        }

        // web: const visible = compact ? entries.slice(0, 4) : entries
        IReadOnlyList<WidgetDetailEntry> visible;
        if (model.Compact && model.Entries.Count > CompactLimit)
        {
            visible = model.Entries.Take(CompactLimit).ToList();
        }
        else
        {
            visible = model.Entries;
        }

        var rows = new List<WidgetDetailRowDisplay>(visible.Count);
        for (int i = 0; i < visible.Count; i++)
        {
            WidgetDetailEntry entry = visible[i];
            string displayValue = entry.Value ?? EmDash;
            bool hasBadge = entry.Badge is not null;
            string badgeText = entry.Badge?.Text ?? string.Empty;
            StatusKind badgeStatus = entry.Badge is { } badge ? StatusFor(badge.Variant) : StatusKind.Neutral;

            rows.Add(new WidgetDetailRowDisplay(
                Label: entry.Label,
                DisplayValue: displayValue,
                Mono: entry.Mono,
                HasBadge: hasBadge,
                BadgeText: badgeText,
                BadgeStatus: badgeStatus,
                ShowDivider: i < visible.Count - 1,
                AutomationName: ComposeName(entry.Label, displayValue, hasBadge ? badgeText : null)));
        }

        return new WidgetDetailCardDisplay(
            IsEmpty: false,
            EmptyMessage: string.Empty,
            EmptyIconGlyph: null,
            Rows: rows);
    }

    /// <summary>Map a web badge variant onto the native status tone (the web <c>badgeVariantMap</c>).</summary>
    /// <param name="variant">The web badge variant.</param>
    public static StatusKind StatusFor(WidgetDetailBadgeVariant variant) => variant switch
    {
        WidgetDetailBadgeVariant.Success => StatusKind.Success,
        WidgetDetailBadgeVariant.Warning => StatusKind.Warning,
        WidgetDetailBadgeVariant.Error => StatusKind.Danger,
        _ => StatusKind.Neutral,
    };

    // Narrator reads each (non-interactive) row as "label: value[, badge]" so the paired label/value/chip are
    // announced together; the web rows have no explicit aria, so this is a native accessibility aid built from
    // the same data, using the original (non-uppercased) label.
    private static string ComposeName(string label, string value, string? badge) =>
        badge is null
            ? $"{label}: {value}"
            : $"{label}: {value}, {badge}";
}

/// <summary>
/// PII-safe diagnostics for the <c>WidgetDetailCard</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never the row labels or values — so a
/// diagnostics line can never leak fleet state. Thread-safe.
/// </summary>
public sealed class WidgetDetailCardDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">An optional sink the operational event line is written to.</param>
    public WidgetDetailCardDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=WidgetDetailCard</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={WidgetDetailCardRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/widget-primitives/WidgetDetailCard/WidgetDetailCardModel.kt
// Pure, framework-free model + projection + a11y affordances + diagnostics for the WidgetDetailCard shared
// widget primitive — the native analogue of every decision the web component makes
// (web/src/features/dashboard/widgets/shared/WidgetDetailCard.tsx) before it paints. No Compose, no Android,
// no HTTP: every declaration here is exercised off-device in the :app:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// What the web source is (and therefore the COMPLETE branch set this surface reproduces):
//   • A PURE, presentational building block reused by many dashboard widgets: it takes a list of
//     `DetailEntry { label, value: string | number | null, badge?, mono? }` plus a `compact` flag, an
//     `emptyMessage` and an `emptyIcon`, and renders a vertically-scrolling stack of label/value rows. There is
//     NO hook, NO fetch and NO data port to bind (no P1/S8 state holder, no Source/ViewModel) — modelling one
//     would invent an async dependency the web spec does not have (honesty covenant: no scope narrowing, no
//     silent drift). The sibling presentational ports Accordion / VisuallyHidden set the same precedent
//     (composable + pure model, no Source/ViewModel).
//   • So the surface's REAL, fully-reproduced states are: the EMPTY state (web `entries.length === 0` → the
//     shared `EmptyState`, never a blank box) and the POPULATED state crossed with its per-row branches — a
//     badge present/absent (web `{entry.badge && …}`), a monospace value vs a proportional one (web
//     `entry.mono && 'font-mono'`), a null value rendered as an em dash (web `entry.value ?? '—'`), the
//     compact cap that shows at most the first four rows (web `compact ? entries.slice(0, 4) : entries`), and
//     the hairline divider under every row except the last (web `i < visible.length - 1 && 'border-b …'`).
//     Each is reduced here in [projectWidgetDetailCard] and asserted off-device, doubling as the per-state
//     snapshot.
//
// Why the generic data-surface states (loading / error / stale / offline) are intentionally absent: this
// primitive fetches nothing — its rows are handed in by the parent widget. There is no query to be loading, to
// fail, to go stale, or to be offline, so inventing those states would be dishonest. The owning widget that DOES
// fetch (and can be loading / stale / offline) renders its own data surface and only hands this card the rows it
// already resolved.
//
// i18n: the web source renders no `t()` key of its own (its labels, values and `emptyMessage` are all
// caller-supplied, already-localized strings); its ONLY built-in string is the hardcoded English default
// `'No details available'`. That default resolves here through the i18n facade by-name ([resolveOptional], the
// native mirror of i18next `t(key, default)`) with the English [WidgetDetailCardDefaults.EMPTY_MESSAGE] fallback
// — exactly the Accordion precedent for a string the web owns implicitly rather than through the catalog — so no
// English literal is hardcoded at the render boundary and a future catalog entry is picked up automatically.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/widget-primitives/WidgetDetailCard — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName`
// is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.widgetprimitives.widgetdetailcard

import io.teslasync.android.components.ui.BadgeVariant
import io.teslasync.shared.core.diagnostics.Logger

/**
 * Diagnostics surface slug emitted with the `view.opened` event (P1/S11). Carries no label, value, or badge —
 * only this constant identifier — so a diagnostics line can never leak what the card shows. Matches the
 * prompt-mandated surface slug (`WidgetDetailCard`).
 */
const val WIDGET_DETAIL_CARD_SLUG: String = "WidgetDetailCard"

/**
 * The semantic tone of a [DetailBadge] — the native tag for the web `badge.variant`
 * (`'success' | 'warning' | 'error' | 'neutral'`). Mapped onto the shared [BadgeVariant] at the render boundary
 * by [badgeVariantFor], the faithful port of the web `badgeVariantMap`.
 */
enum class DetailBadgeVariant { Success, Warning, Error, Neutral }

/**
 * An optional status chip rendered after a row's value — the native analogue of the web
 * `badge?: { text; variant }`. [text] is a caller-supplied, already-localized short label; [variant] selects the
 * chip's tone.
 */
data class DetailBadge(
    val text: String,
    val variant: DetailBadgeVariant,
)

/**
 * One label/value row handed in by the parent widget — the native analogue of the web `DetailEntry`.
 *
 * @property label the row's caller-supplied, already-localized label (web `label`); shown uppercased + muted.
 * @property value the row's already-formatted display string (web `value`); `null` renders as an em dash
 *   ([EM_DASH]), exactly as the web `entry.value ?? '—'`. Numbers are formatted to a string at
 *   the call site (the native convention), mirroring React's `string | number` coercion.
 * @property badge an optional trailing status chip (web `badge`); `null` ⇒ no chip.
 * @property mono renders the value in a monospace face when `true` (web `entry.mono && 'font-mono'`).
 */
data class DetailEntry(
    val label: String,
    val value: String?,
    val badge: DetailBadge? = null,
    val mono: Boolean = false,
)

/** The em dash shown for a null value — the native mirror of the web `entry.value ?? '—'`. */
const val EM_DASH: String = "\u2014"

/** How many rows the compact variant shows at most — the native mirror of web `entries.slice(0, 4)`. */
const val COMPACT_ROW_LIMIT: Int = 4

/**
 * A render-ready row reduced from a [DetailEntry] — everything the view needs to draw one line, so the value
 * fallback, the divider rule and the slot flags are decided once here and unit-tested off-device.
 *
 * @property label the (original-case) row label; the view uppercases it for display while a11y reads this.
 * @property value the resolved display value (web `value ?? '—'`); never null.
 * @property badge the optional trailing chip, carried through verbatim.
 * @property mono whether the value renders monospace.
 * @property showDivider whether a hairline divider is drawn beneath the row — every row except the last
 *   (web `i < visible.length - 1`).
 */
data class DetailRow(
    val label: String,
    val value: String,
    val badge: DetailBadge?,
    val mono: Boolean,
    val showDivider: Boolean,
)

/**
 * The render-ready classification of the whole card — either the empty branch (web `entries.length === 0`) or
 * the populated list of [rows]. The card always renders SOMETHING (rows or the friendly empty state), so there
 * is no hidden surface.
 *
 * @property isEmpty no rows were supplied — the view shows the shared empty state, never a blank box.
 * @property rows the render-ready rows in order (empty when [isEmpty]).
 */
data class WidgetDetailCardRender(
    val isEmpty: Boolean,
    val rows: List<DetailRow>,
)

/**
 * Resolve a row's display value — the native mirror of the web `entry.value ?? '—'`. A non-null value
 * (including the empty string and `"0"`) is shown verbatim; only `null` becomes the em dash.
 */
fun resolveDetailValue(value: String?): String = value ?: EM_DASH

/**
 * The rows actually drawn for [entries] — all of them, or the first [COMPACT_ROW_LIMIT] when [compact] is set
 * (web `compact ? entries.slice(0, 4) : entries`). `take` is safe when fewer than the cap exist.
 */
fun visibleDetailEntries(
    entries: List<DetailEntry>,
    compact: Boolean,
): List<DetailEntry> = if (compact) entries.take(COMPACT_ROW_LIMIT) else entries

/**
 * Reduce the parent's [entries] into the render-ready [WidgetDetailCardRender]. Pure (no Compose). An empty
 * list flags [WidgetDetailCardRender.isEmpty] so the view shows the shared empty state (the prompt's
 * "empty → friendly empty state, never a blank box" contract); otherwise each visible entry becomes a
 * [DetailRow] with its value resolved and its divider flagged for every row but the last.
 */
fun projectWidgetDetailCard(
    entries: List<DetailEntry>,
    compact: Boolean,
): WidgetDetailCardRender {
    if (entries.isEmpty()) {
        return WidgetDetailCardRender(isEmpty = true, rows = emptyList())
    }
    val visible = visibleDetailEntries(entries, compact)
    val lastIndex = visible.size - 1
    val rows =
        visible.mapIndexed { index, entry ->
            DetailRow(
                label = entry.label,
                value = resolveDetailValue(entry.value),
                badge = entry.badge,
                mono = entry.mono,
                showDivider = index < lastIndex,
            )
        }
    return WidgetDetailCardRender(isEmpty = false, rows = rows)
}

/**
 * Map a [DetailBadgeVariant] onto the shared [BadgeVariant] — the faithful port of the web `badgeVariantMap`
 * (`success→success, warning→warning, error→danger, neutral→neutral`). Kept pure so the mapping is unit-tested
 * without a Compose host; `error` deliberately lands on [BadgeVariant.Danger] (the shared chip's tone name).
 */
fun badgeVariantFor(variant: DetailBadgeVariant): BadgeVariant =
    when (variant) {
        DetailBadgeVariant.Success -> BadgeVariant.Success
        DetailBadgeVariant.Warning -> BadgeVariant.Warning
        DetailBadgeVariant.Error -> BadgeVariant.Danger
        DetailBadgeVariant.Neutral -> BadgeVariant.Neutral
    }

/**
 * The single spoken description for a row — so TalkBack reads "label, value[, badge]" as one unit instead of
 * spelling out the visually-uppercased label letter by letter. Built from the ORIGINAL-case [DetailRow.label]
 * and the resolved value (and badge text when present), it is applied at the render boundary with
 * `clearAndSetSemantics`. Pure so the a11y label is unit-tested off-device.
 */
fun detailRowContentDescription(row: DetailRow): String =
    buildString {
        append(row.label)
        append(": ")
        append(row.value)
        row.badge?.let { badge ->
            append(", ")
            append(badge.text)
        }
    }

/**
 * Native-only default copy — the web source's one built-in string. Absent a catalog hit this English fallback is
 * used (the native mirror of i18next's default argument), matching the Accordion precedent for a string the web
 * owns implicitly rather than through a `t()` key.
 */
object WidgetDetailCardDefaults {
    /** The default empty-state message — web `emptyMessage ?? 'No details available'`. */
    const val EMPTY_MESSAGE: String = "No details available"
}

/** Resource name for the default empty-state message (by-name; absent ⇒ the English [WidgetDetailCardDefaults.EMPTY_MESSAGE]). */
const val KEY_WIDGET_DETAIL_CARD_EMPTY: String = "translation_widget_detailCard_empty"

/**
 * Reproduces i18next's `t(key, default)` against the native i18n facade: returns the [lookup] result for
 * [resourceName] when it resolves to a non-blank string, otherwise the [fallback] default. [lookup] is a thin
 * seam over the Android string catalog in production (an optional by-name resource read) and a map in tests, so
 * the resolve-or-fallback decision stays pure and unit-tested.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never a label,
 * value, or badge — so a diagnostics line can never leak what the card shows.
 */
object WidgetDetailCardDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = WIDGET_DETAIL_CARD_SLUG

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
